src/Microsoft.AspNetCore.Authorization.Policy/AuthenticationPolicyEvaluator.cs

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Implements <see cref="IAuthenticationPolicyEvaluator"/>.
+    /// </summary>
+    public class AuthenticationPolicyEvaluator : IAuthenticationPolicyEvaluator
+    {
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="policyProvider">The The <see cref="IAuthenticationPolicyProvider"/>.</param>
+        public AuthenticationPolicyEvaluator(IAuthenticationPolicyProvider policyProvider)
+        {
+            Policies = policyProvider;
+        }
+
+        /// <summary>
+        /// Used to lookup AuthenticationPolicies.
+        /// </summary>
+        public IAuthenticationPolicyProvider Policies { get; }
+
+        /// <summary>
+        /// Authenticate for the specified authentication policy.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <param name="policyName">The name of the authentication policy.</param>
+        /// <returns>The result.</returns>
+        public virtual async Task<AuthenticateResult> AuthenticateAsync(HttpContext context, string policyName)
+        {
+            var policy = await Policies.GetAsync(context, policyName);
+            if (policy != null)
+            {
+                var count = policy.AuthenticateSchemes.Count();
+                if (count > 0)
+                {
+                    ClaimsPrincipal principal = null;
+                    foreach (var s in policy.AuthenticateSchemes)
+                    {
+                        var auth = await context.AuthenticateAsync(s);
+                        if (auth.Succeeded)
+                        {
+                            if (principal == null)
+                            {
+                                principal = auth.Principal;
+                            }
+                            else
+                            {
+                                // REVIEW: what about the auth properties, we just dropped them...
```

Comment:
How did this work before? Did it ever?

Reply:
Nope, never worked, whenever we merged principals we would lose whatever was in the auth properties, but to be fair, you couldn't access the properties in 1.x unless you called the appropriate Authenticate overload anyways